include/net/net_pkt.h

```diff
@@ -1998,6 +2001,158 @@ const char *net_pkt_pool2str(struct net_buf_pool *pool);
 #define net_pkt_print(...)
 #endif /* CONFIG_NET_PKT_LOG_LEVEL >= LOG_LEVEL_DBG */
 
+/* New allocator, and API are defined below.
+ * This will be simpler when time will come to get rid of former API above.
+ */
+#if defined(CONFIG_NET_DEBUG_NET_PKT_ALLOC) || \
+	(CONFIG_NET_PKT_LOG_LEVEL >= LOG_LEVEL_DBG)
+
+struct net_pkt *net_pkt_alloc_debug(s32_t timeout,
+				    const char *caller, int line);
+#define net_pkt_alloc(_timeout)					\
+	net_pkt_alloc_debug(_timeout, __func__, __LINE__)
+
+struct net_pkt *net_pkt_rx_alloc_debug(s32_t timeout,
+				       const char *caller, int line);
+#define net_pkt_rx_alloc(_timeout)				\
+	net_pkt_rx_alloc_debug(_timeout, __func__, __LINE__)
+
+struct net_pkt *net_pkt_alloc_on_iface_debug(struct net_if *iface,
+					     s32_t timeout,
+					     const char *caller,
+					     int line);
+#define net_pkt_alloc_on_iface_(_iface, _timeout)			\
+	net_pkt_alloc_on_iface_debug(_iface, _timeout, __func__, __LINE__)
+
+struct net_pkt *net_pkt_rx_alloc_on_iface_debug(struct net_if *iface,
+						s32_t timeout,
+						const char *caller,
+						int line);
+#define net_pkt_rx_alloc_on_iface_(_iface, _timeout)			\
+	net_pkt_rx_alloc_on_iface_debug(_iface, _timeout,		\
+					   __func__, __LINE__)
+
+int net_pkt_alloc_buffer_debug(struct net_pkt *pkt,
+			       size_t size,
+			       enum net_ip_protocol proto,
+			       s32_t timeout,
+			       const char *caller, int line);
+#define net_pkt_alloc_buffer(_pkt, _size, _proto, _timeout)		\
+	net_pkt_alloc_buffer_debug(_pkt, _size, _proto, _timeout,	\
+				      __func__, __LINE__)
+
+struct net_pkt *net_pkt_alloc_with_buffer_debug(struct net_if *iface,
+						size_t size,
+						sa_family_t family,
+						enum net_ip_protocol proto,
+						s32_t timeout,
+						const char *caller,
+						int line);
+#define net_pkt_alloc_with_buffer(_iface, _size, _family,		\
+				  _proto, _timeout)			\
+	net_pkt_alloc_with_buffer_debug(_iface, _size, _family,	\
+					   _proto, _timeout,		\
+					   __func__, __LINE__)
+
+struct net_pkt *net_pkt_rx_alloc_with_buffer_debug(struct net_if *iface,
+						   size_t size,
+						   sa_family_t family,
+						   enum net_ip_protocol proto,
+						   s32_t timeout,
+						   const char *caller,
+						   int line);
+#define net_pkt_rx_alloc_with_buffer(_iface, _size, _family,		\
+				     _proto, _timeout)			\
+	net_pkt_rx_alloc_with_buffer_debug(_iface, _size, _family,	\
+					      _proto, _timeout,		\
+					      __func__, __LINE__)
+#else
+
+/**
+ * @brief Allocate an initialized net_pkt
+ *
+ * Note: for the time being, 2 pools are used. One for TX and one for RX.
+ *       This allocater has to be used for TX.
```

Comment:
So, why it's not named `net_pkt_tc_alloc`?

Reply:
This "rx/tx pools" is not a good idea, it was a bandage on a wooden leg when it first appeared (worst one being the net_context dedicated pool). I want to get rid of it. So here, the core itself will not use any "tx" wording anywhere. For the time being, only drivers will have the differentiation with "rx".

Then one I am done with the API, I'll get rid of the tx/rx pools, one bit in net_pkt will tell the direction. unset: tx, set: rx. and rx will be set in net_recv_data(), then in memory info we will still know on what path packet are allocated. And that's all, that will be enough for what this differentiation is really used (and no such differentiation has never been a proper fix for memory leak or else)

Reply:
So, later is later, in the meantime, this patch gives "net_pkt_rx_alloc" vs "net_pkt_alloc", which is weird.

And using packet pools for different purposes, for example for RX vs TX direction, is the best practice to write reliable applications, so it needs to be supported anyway.

Reply:
> is the best practice to write reliable applications

I don't know how you ended up to such conclusion.
If either rx or tx pool are empty due to a leak, the result will be _exactly_ the same as with one pool: there is nothing that can be done to recover.
Now in case one wants to limit how greedy is the rx path over tx path or vice versa, that also can be done with one pool, or limiting the rx/tx fifos etc... there are various ways. But having 2 pools is not the unique solution for this. Plus, the downside of this 2+ pools scheme is also the code everywhere: you need to use the right allocator, thus you need dedicated functions, thus it eats up a bit more ram/rom etc etc... It's only more burden for no good reasons.